engine/src/main/java/org/terasology/rendering/world/dag/ShadowMapNode.java

```diff
@@ -0,0 +1,171 @@
+    // FIXME: unnecessary arguments must be eliminated
+    public ShadowMapNode(Material shader,
+                         Camera playerCamera,
+                         WorldRendererImpl worldRenderer,
+                         BackdropProvider backdropProvider,
+                         RenderingConfig renderingConfig,
+                         GraphicState state) {
+
+        this.shader = shader;
+        this.playerCamera = playerCamera;
+        this.worldRenderer = worldRenderer;
+        this.backdropProvider = backdropProvider;
+        this.renderingConfig = renderingConfig;
+        this.state = state;
+
+    }
+
+
+    @Override
+    public void initialise() {
+        // TODO: shader and fbo shall be initialised here
+        // TODO: using ShaderParametersShadowMap here, instead of storing in ShaderManagerLwjgl
+    }
+
+    @Override
+    public void preRender(float update) {
+        // positionShadowMapCamera()
+
+        // Shadows are rendered around the player so...
+        Vector3f lightPosition = new Vector3f(playerCamera.getPosition().x, 0.0f, playerCamera.getPosition().z);
+
+        // Project the shadowMapCamera position to light space and make sure it is only moved in texel steps (avoids flickering when moving the shadowMapCamera)
+        float texelSize = 1.0f / renderingConfig.getShadowMapResolution();
+        texelSize *= 2.0f;
+
+        camera.getViewProjectionMatrix().transformPoint(lightPosition);
+        lightPosition.set(TeraMath.fastFloor(lightPosition.x / texelSize) * texelSize, 0.0f, TeraMath.fastFloor(lightPosition.z / texelSize) * texelSize);
+        camera.getInverseViewProjectionMatrix().transformPoint(lightPosition);
+
+        // ... we position our new shadowMapCamera at the position of the player and move it
+        // quite a bit into the direction of the sun (our main light).
+
+        // Make sure the sun does not move too often since it causes massive shadow flickering (from hell to the max)!
+        float stepSize = 50f;
+        Vector3f sunDirection = backdropProvider.getQuantizedSunDirection(stepSize);
+
+        Vector3f sunPosition = new Vector3f(sunDirection);
+        sunPosition.scale(256.0f + 64.0f);
+        lightPosition.add(sunPosition);
+
+        camera.getPosition().set(lightPosition);
+
+        // and adjust it to look from the sun direction into the direction of our player
+        Vector3f negSunDirection = new Vector3f(sunDirection);
+        negSunDirection.scale(-1.0f);
+
+        camera.getViewingDirection().set(negSunDirection);
+
+        camera.update(update);
+
+    }
+
+
+    @Override
+    public void render() {
+        // TODO: removing this assignment
+        fbo = state.buffers.sceneShadowMap; // added here since graphic settings can be modified and a new fbo can be generated.
+
+        // preRenderSetupSceneShadowMap
+        fbo.bind();
+        glViewport(0, 0, fbo.width(), fbo.height());
+        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
```

Comment:
Please add a TODO here: verify the need to clear color buffer - if I remember correctly the shadowMap fbo has no color buffer at all!